### PR TITLE
Feature: filename Subcommand

### PR DESCRIPTION
**Pull request by BJTMastermind**
This pull request introduces a new `filename` subcommand to `/audioplayer`.

### Overview

It operates by utilizing a new file named `filename_mappings.dat`, which resides in the existing `audio_player_data` folder of the world. This file is structured as a GZipped NBT file, containing a list of mappings between original filenames and their corresponding UUID names.

### File Structure

The structure of `filename_mappings.dat` is as follows:

```
TAG_Compound(): {
    TAG_List("mappings"): {
        TAG_Compound(): {
            TAG_String("filename"): "<Original filename>.<extension>",
            TAG_String("uuid"): "<UUID>.<extension>"
        },
        ...
    }
}
```

### Functionality

Whenever a new audio file is uploaded, regardless of the method used, an entry is added to `filename_mappings.dat`. Conversely, if a file is deleted the mappings associated with that file will be removed the next time the `filename` command is ran or when another audio file is uploaded.

### Motivation

I implemented this feature to enhance the usability and maintainability of the users audio files. By introducing the `filename` subcommand and associated functionality, users can easily manage audio file mappings, whether its with the command itself for a couple of files or externally with user made scripts for a larger number of files.

### CODE REVIEW

**Comment by BJTMastermind** [comment]
This should be good now.

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            CompoundTag root = loadOrDefault(mappingsFile);
+            ListTag mappings = getMappingListOrDefault(root);
+
+            CompoundTag entry = new CompoundTag();
+            entry.putString("uuid", sound.toString() + extension);
+            entry.putString("filename", filename);
+
+            mappings.add(entry);
+
+            if (!root.contains("mappings")) {
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:31]: Why does the root compound need a nested list? Couldn't the root be the list?

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                root.put("mappings", mappings);
+            }
+
+            NbtIo.writeCompressed(root, mappingsFile.toPath());
+        } catch (IOException e) {
+            System.err.println("Permission Denied. Failed to read or write the file.");
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:37]: `Denied` shouldn't be capitalized

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        verifyFilesExist(server);
+    }
+
+    public static void verifyFilesExist(MinecraftServer server) {
+        URI location = server.getWorldPath(AudioManager.AUDIO_DATA).resolve("filename_mappings.dat").toUri();
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:44]: Why are you using a URI? Use a path with .toFile()

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            if (mappings.isEmpty()) {
+                mappingsFile.delete();
+            }
+        } catch (IOException e) {
+            System.err.println("Permission Denied. Failed to read or write the file.");
+        }
+    }
+
+    public static void autoSaveHappened() {
+        AudioPlayer.LOGGER.warn("Auto Save Triggered!");
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:71]: Why is this needed??

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    private static CompoundTag loadOrDefault(File filepath) throws IOException {
+        if (filepath.exists()) {
+            return NbtIo.readCompressed(filepath.toPath(), NbtAccounter.unlimitedHeap());
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:76]: Why `NbtAccounter.unlimitedHeap()`?

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import java.net.URI;
+import java.util.UUID;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.NbtAccounter;
+import net.minecraft.nbt.NbtIo;
+import net.minecraft.server.MinecraftServer;
+
+public class FilenameMappings {
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:14]: All string fields should be constants

Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:14]: It is VERY inefficient reading the whole file every time you want to read or save a mapping. Either use caching or keep it in memory and make `FilenameMappings` non-static.

File: src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java [hunk 7]
```diff
@@ -75,4 +88,67 @@ public void id(CommandContext<CommandSourceStack> context) throws CommandSyntaxE
...
+        if (customSound == null) {
+            context.getSource().sendFailure(Component.literal("Item does not have custom audio"));
+            return;
+        }
+
+        String soundId = customSound.getSoundId().toString();
+
+        MinecraftServer server = player.getCommandSenderWorld().getServer();
+
+        File mappingsFile = new File(server.getWorldPath(AudioManager.AUDIO_DATA).resolve("filename_mappings.dat").toUri());
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java:113]: This is redundant logic!

**Review by henkelmax — COMMENTED** [review]
To be honest, this is a mess...
I might be better off implementing this myself. Reviewing this might have taken me longer than actually implementing it properly myself.

**Comment by henkelmax** [comment]
Also please explain me what's the purpose of this command?
It basically just allows you to get the file name of a custom sound item. It can't be used anywhere, so I don't really see a point of this at all.

**Comment by henkelmax** [comment]
35883d6c4b50fd0d6a2a8d5f0067084ed7aab980 should properly implement file names that can be actually used for applying them to items.

**Comment by BJTMastermind** [comment]
I apologize that the code is a little messy. 

This is the first time im contributing to a project so im sure there's a lot for me to learn in making better contributions and code in general. 

The main feature from this I would have used is the nbt file database as I was recently converting a world that used this mod for custom audio to bedrock edition and since all the sounds are just uuids it makes it hard to know whats what, so the database of mappings would allow me to setup a script to rename them to their original uploaded name to better know what the files are just by the name, this is most helpful when there are a lot of audio files.

I added the `filename` command along with it for when there are only a few audio files the command could be used to copy the original name and rename the file based on it id from the `id` command. but I can see from an in game point of view this command doesn't do much.

I thought about maybe having the file name be the default display if you don't give the item one but i figured that some might want it to be blank so i didn't end up do this.

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java (hunk 5 above)
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:76]: I didn't want to use `unlimitedHeap()` but i wasn't sure what to put as the size for the other method for that.

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java (hunk 4 above)
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:71]: Oh, sorry thats not. I must of forgot to remove that, I was tested to see if i could auto check the mappings when a auto save happened but i couldn't find a way to detected that event.

File: src/main/java/de/maxhenkel/audioplayer/Filebin.java [hunk 8]
```diff
@@ -62,6 +62,7 @@ public static void downloadSound(MinecraftServer server, UUID sound) throws IOEx
 
                 String filename = file.get("filename").getAsString();
                 AudioManager.saveSound(server, sound, url + "/" + filename);
+                FilenameMappings.append(server, filename, sound);
```
[the remark this answers is not in the document]
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/Filebin.java:65]: I looked for a centralized upload area of the code but i didn't find anything, unless I overlooked something. which is why i did it at each upload method.

File: src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java [hunk 9]
```diff
@@ -75,4 +88,67 @@ public void id(CommandContext<CommandSourceStack> context) throws CommandSyntaxE
...
+
+        if (!mappingsFile.exists()) {
+            context.getSource().sendFailure(Component.literal(
+                "Item has custom audio but filename was not saved when it was created. "+
```
[the remark this answers is not in the document]
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java:117]: Only because the line got really long. I don't think it would affect how it appears in game though.

File: src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java [hunk 10]
```diff
@@ -75,4 +88,67 @@ public void id(CommandContext<CommandSourceStack> context) throws CommandSyntaxE
...
+            );
+            return;
+        }
+
+        CompoundTag root = NbtIo.readCompressed(mappingsFile.toPath(), NbtAccounter.unlimitedHeap());
+        ListTag mappings = (ListTag) root.get("mappings");
+
+        for (int i = 0; i < mappings.size(); i++) {
+            CompoundTag entry = mappings.getCompound(i);
+            String uuid = entry.getString("uuid").replaceAll(".wav|.mp3", "");
```
[the remark this answers is not in the document]
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java:128]: I think its unlikely that 2 files of different types have the same uuid name but i suppose in this case it would just return the first one it finds.

File: src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java [hunk 11]
```diff
@@ -75,4 +88,67 @@ public void id(CommandContext<CommandSourceStack> context) throws CommandSyntaxE
...
+                    );
+
+                context.getSource().sendSuccess(() -> msg, false);
+                FilenameMappings.verifyFilesExist(server);
+                return;
+            }
+        }
+
+        context.getSource().sendFailure(Component.literal(
+            "Failed to retrieved sound filename.\n"+
```
[the remark this answers is not in the document]
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/command/UtilityCommands.java:150]: Sorry, typo

File: src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java (hunk 1 above)
Comment (BJTMastermind) [reply src/main/java/de/maxhenkel/audioplayer/FilenameMappings.java:31]: This might be a oversight, In the nbt library i normally use for projects you have to use a compound tag as the root so thats why i did it that way, force of habit. but maybe Mojangs inbuilt library allows a list to be the root.

**Comment by henkelmax** [comment]
> I apologize that the code is a little messy.

> This is the first time im contributing to a project so im sure there's a lot for me to learn in making better contributions and code in general.

No worries.

Basically how I implemented it, is pretty similar. The file names are stored in a json, which is also more human readable as NBT. Theres also a command that gives you the file name of an item and additionally lets you apply audio to an item by file name.

**Comment by BJTMastermind** [comment]
I was just checking it out. I noticed the apply command that pops up when you click `[Put on item]` might be broken a bit as it was given me the error `No audio with name '<uuid>' found or more then one found` but when I used the musicdisc command instead I was able to apply the sound to the disc.

**Comment by BJTMastermind** [comment]
Yeah its on the latest commit. I could make a bug report to give more detail about it and maybe a short clip of running the commands. If thats easier for you.

**Comment by henkelmax** [comment]
You can post that here. I'll also do some testing.

**Comment by henkelmax** [comment]
Thanks, that might be an issue that doesn't happen in my development environment.

**Comment by henkelmax** [comment]
Can you try if it works for you now?

**Comment by BJTMastermind** [comment]
Looks to be working now.

**Comment by henkelmax** [comment]
Alright, thanks for testing that out!
